src/core/openjpeg_decoder.js

```diff
@@ -0,0 +1,102 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BaseException } from "../shared/util.js";
+import OpenJPEG from "../../external/openjpeg/openjpeg.js";
+
+class JpxError extends BaseException {
+  constructor(msg) {
+    super(`JPX error: ${msg}`, "JpxError");
+  }
+}
+
+class OpenJPEGDecoder {
+  #module = null;
+
+  static #instance = null;
+
+  constructor(module) {
+    this.#module = module;
+  }
+
+  static decode(data) {
+    this.#instance ||= new OpenJPEGDecoder(OpenJPEG());
+    return this.#instance.#decode(data);
+  }
+
+  #decode(data) {
+    const mod = this.#module;
+    const data_ptr = mod._malloc(data.length);
+    const image_ptr_ptr = mod._malloc(4);
+    const image_size_ptr = mod._malloc(4);
+
+    mod.HEAPU8.set(data, data_ptr);
+    const ret = mod._jp2_decode(
+      data_ptr,
+      data.length,
+      image_ptr_ptr,
+      image_size_ptr
+    );
+    mod._free(data_ptr);
+    if (ret) {
+      mod._free(image_ptr_ptr);
+      mod._free(image_size_ptr);
+      return new Uint8Array(0);
+    }
+
+    const image_ptr = mod.HEAPU32[image_ptr_ptr / 4];
+    const image_size = mod.HEAPU32[image_size_ptr / 4];
+    const newArray = mod.HEAPU8.slice(image_ptr, image_ptr + image_size);
+
+    mod._free(image_ptr);
+    mod._free(image_ptr_ptr);
+    mod._free(image_size_ptr);
+
+    return newArray;
+  }
+
+  static parseImageProperties(stream) {
+    let newByte = stream.getByte();
+    const properties = {
+      width: 0,
+      height: 0,
+      // Results are always returned as `Uint8ClampedArray`s.
+      bitsPerComponent: 8,
+      componentsCount: 0,
+    };
+    while (newByte >= 0) {
+      const oldByte = newByte;
+      newByte = stream.getByte();
+      const code = (oldByte << 8) | newByte;
+      // Image and tile size (SIZ)
+      if (code === 0xff51) {
+        stream.skip(4);
+        const Xsiz = stream.getInt32() >>> 0; // Byte 4
+        const Ysiz = stream.getInt32() >>> 0; // Byte 8
+        const XOsiz = stream.getInt32() >>> 0; // Byte 12
+        const YOsiz = stream.getInt32() >>> 0; // Byte 16
+        stream.skip(16);
+        const Csiz = stream.getUint16(); // Byte 36
+        properties.width = Xsiz - XOsiz;
+        properties.height = Ysiz - YOsiz;
+        properties.componentsCount = Csiz;
+        return properties;
+      }
+    }
+    throw new JpxError("No size marker found in JPX stream");
+  }
+}
+
+export { JpxError, OpenJPEGDecoder };
```

Comment:
The `JpxError` exception doesn't seem to actually be caught anywhere since before it was also only used in `jpx.js` which is now removed. While it makes sense to raise a specific exception here internally in this file, I don't think we need to export it here?

Note that the fuzzers only use the exception message, see https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L10-L14, so they also don't need the exception class itself, and the image decoders bundle also didn't expose them. However, depending on what we decide to do with the JPX fuzzer in the other comment, we might need to change the fuzzer implementation at https://github.com/mozilla/pdf.js/blob/master/test/fuzz/jpx_image.fuzz.js#L10 because I can imagine that the "Cannot read properties" message might not be emitted by OpenJPEG anymore and something else is logged instead.